OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,460 @@
+  static constexpr uint32_t kLayerSpecVersion = VK_API_VERSION_1_1;
+
+  static constexpr std::array<VkExtensionProperties, 3> kDeviceExtensions = {
+      VkExtensionProperties{.extensionName = VK_EXT_DEBUG_MARKER_EXTENSION_NAME,
+                            .specVersion = VK_EXT_DEBUG_MARKER_SPEC_VERSION},
+      VkExtensionProperties{.extensionName = VK_EXT_DEBUG_UTILS_EXTENSION_NAME,
+                            .specVersion = VK_EXT_DEBUG_UTILS_SPEC_VERSION},
+      VkExtensionProperties{.extensionName = VK_EXT_HOST_QUERY_RESET_EXTENSION_NAME,
+                            .specVersion = VK_EXT_HOST_QUERY_RESET_SPEC_VERSION}};
+
+  VulkanLayerController()
+      : device_manager_(&dispatch_table_),
+        timer_query_pool_(&dispatch_table_, kNumTimerQuerySlots),
+        submission_tracker_(&dispatch_table_, &timer_query_pool_, &device_manager_,
+                            std::numeric_limits<uint32_t>::max()) {}
+
+  ~VulkanLayerController() { CloseVulkanLayerProducerIfNecessary(); }
+
+  // ----------------------------------------------------------------------------
+  // Layer bootstrapping code
+  // ----------------------------------------------------------------------------
+
+  [[nodiscard]] VkResult OnCreateInstance(const VkInstanceCreateInfo* create_info,
+                                          const VkAllocationCallbacks* allocator,
+                                          VkInstance* instance) {
+    // The specification ensures that the create_info pointer is not nullptr.
+    CHECK(create_info != nullptr);
+
+    auto* layer_create_info = absl::bit_cast<VkLayerInstanceCreateInfo*>(create_info->pNext);
+
+    // Iterate over the create info chain to find the layer linkage information. This contains
+    // the GetInstanceProcAddr function of the next layer (or the driver if this is the last layer).
+    while (layer_create_info != nullptr &&
+           (layer_create_info->sType != VK_STRUCTURE_TYPE_LOADER_INSTANCE_CREATE_INFO ||
+            layer_create_info->function != VK_LAYER_LINK_INFO)) {
+      layer_create_info = absl::bit_cast<VkLayerInstanceCreateInfo*>(layer_create_info->pNext);
+    }
+
+    if (layer_create_info == nullptr) {
+      return VK_ERROR_INITIALIZATION_FAILED;
+    }
+
+    InitVulkanLayerProducerIfNecessary();
+
+    PFN_vkGetInstanceProcAddr next_get_instance_proc_addr_function =
+        layer_create_info->u.pLayerInfo->pfnNextGetInstanceProcAddr;
+
+    // Advance linkage for next layer.
+    layer_create_info->u.pLayerInfo = layer_create_info->u.pLayerInfo->pNext;
+
+    // Need to call vkCreateInstance down the chain to actually create the
+    // instance, as we need the instance to be alive in the create instance dispatch table.
+    auto create_instance = absl::bit_cast<PFN_vkCreateInstance>(
+        next_get_instance_proc_addr_function(VK_NULL_HANDLE, "vkCreateInstance"));
+    VkResult result = create_instance(create_info, allocator, instance);
```

Comment:
The comment above seems to suggest that this call must not fail before calling "CreateInstanceDispatchTable". Should we check the result here to make sure sure that it succeeded?

Reply:
Agree we should not fail on error (with CHECK) and actually return the error, but that we also should not do further calls in case of an error. Change with "if" LGTM.